### PR TITLE
#2366: support variables and expressions in repository properties

### LINKED ISSUE #2366: Support for variables and expressions in repository URL
### Feature idea

As an IDEasy maintainer, I would like to be able to configure the fork of the current user as repository.
Since the `git_url` is containing the name of the github user I would like to make use of variables and expressions in a git URL.
Then combining this feature together with #2168 and #989 / #2165 we can then get what we need:
IDEasy.properties:
```
path=IDEasy
git_url=https://github.com/@ask-variable("GIT_USER", "Please enter your github login:", "home", "$[USER]")/IDEasy.git
git_remote=upstream:https://github.com/devonfw/ide-settings.git
git_branch=main
eclipse=import
active=true
```

### Additional context

Feature #2168 is already available but PR #2282 needs to be merged before this story can be implemented.

### PR DESCRIPTION
**Pull request by quando632**
### Implemented changes:

* Every value of a repository `.properties` file is now resolved via the `EnvironmentVariables` hierarchy, so [variables](https://github.com/devonfw/IDEasy/blob/main/documentation/variables.adoc) (`$[VAR]`) and expressions (`@function(...)`) can be used in `git_url`, `git_remote`, `path` and every other property.
* Resolution is done centrally in `RepositoryProperties.doGetProperty(..)` and happens **before** any validation, so it also works for validated properties such as `path` (whose `PATH_PATTERN` would otherwise reject the `[` and `]` of an unresolved `$[VAR]`).
* Legacy support is disabled on purpose (`legacySupport=false`), so the `${...}` syntax of e.g. maven properties in a `build_cmd` is passed through untouched.
* The `active` flag is now evaluated **before** the remaining properties are read, so an expression like `@ask-variable` never asks the user for a repository that is skipped anyway. For this, `RepositoryConfig.loadProperties(RepositoryProperties)` was added as an overload; the existing `loadProperties(Path, IdeContext)` is kept and delegates to it.
* Documented the new capability in `documentation/repository.adoc` and added the CHANGELOG entry.
* Resolved values are cached per `RepositoryProperties` instance. Resolving can ask the user, so a property read twice would ask twice - the `active` flag is read by `RepositoryCommandlet` and again by `RepositoryConfig`.
* When iterating all repositories, a properties file whose expression is malformed is logged and skipped instead of aborting the setup of every other repository. `CliAbortException` is rethrown so an abort by the user is never swallowed; an explicitly requested single repository still reports the failure.

This is the last building block for the use-case described in #2366 - configuring the fork of the current user as repository:

```properties
path=IDEasy
git_url=https://github.com/@ask-variable("GIT_USER", "Please enter your github login:", "home", "$[USERNAME]")/IDEasy.git
git_remote=upstream:https://github.com/devonfw/IDEasy.git
git_branch=main
import=eclipse
active=true
```

> [!NOTE]
> #989 / PR #2282 is merged, `main` has been merged into this branch, and the `@ask-variable` example above was verified end-to-end (see testing instructions). This PR needs no expression specific code itself: #2282 registers the `ExpressionParser` inside `AbstractEnvironmentVariables.resolveRecursive`, i.e. *below* `EnvironmentVariables.resolve(..)`, which is what this PR calls.

> [!NOTE]
> **Known limitation, deliberately not addressed here.** All properties of an active repository are resolved in the first loop of `doRun`, while the "repository already exists / was already setup" checks happen later in `doImportRepository`. A `@ask-secret(..)` in a `git_url` is never persisted by design, so it would be asked on every run even for a repository that is then skipped. `@ask-variable(..)` - the use-case of this issue - persists its answer and is therefore asked only once. Fixing this would require resolving the URL lazily and thus restructuring `RepositoryConfig`, which is out of scope for this story.

### Testing instructions

1. Create or edit a repository properties file in your settings, e.g. `settings/repositories/test.properties`:
   ```properties
   path=repos/$[GIT_USER]
   git_url=https://github.com/$[GIT_USER]/IDEasy.git
   git_remote=upstream:https://github.com/devonfw/IDEasy.git
   active=true
   ```
2. Define the variable, e.g. in `$IDE_HOME/conf/ide.properties`: `GIT_USER=«your-github-login»`
3. Run `ide repository setup test` (or `ide update`). The repository is cloned from `https://github.com/«your-github-login»/IDEasy.git` into `workspaces/main/repos/«your-github-login»`, and `git remote -v` inside the clone shows the additional `upstream` remote.
4. Now remove `GIT_USER` from `ide.properties` and run again: a warning `Undefined variable $[GIT_USER] in '«...»/test.properties'` is logged and the placeholder is kept, so the subsequent `git clone` fails with the unresolved URL visible in the error. An undefined variable in a repository URL is a configuration error and is reported as such rather than being silently swallowed.
5. Set `active=false` and run `ide repository` (without `--force-repositories`): only `Skipping repository test because it is not active, ...` is logged - no resolution and thus no question is triggered for a repository that is skipped.
6. Verify that the legacy `${...}` syntax is *not* resolved - IDEasy must pass it through untouched, since it is common in build commands (maven properties). Use a variable that **is defined** (e.g. the `GIT_USER` from step 2); an undefined one would stay unresolved either way and would prove nothing. Put both syntaxes side by side:
   ```properties
   build_cmd=mvn -Downer=${GIT_USER} -Duser=$[GIT_USER] clean install
   ```
   Run the setup and look at the step that is logged *before* the build starts:
   ```
   Start: Build repository via: mvn -Downer=${GIT_USER} -Duser=quando632 clean install
   ```
   Only the `$[..]` form is replaced, `${..}` stays literal. Note this is about the value IDEasy passes on - the `.properties` file itself is never modified by IDEasy.
   If you would rather not trigger a real maven build, use `path=repos/${GIT_USER}` instead: the log then shows `Invalid path repos/${GIT_USER} from property path of «...»` (the path pattern rejects `{`/`}`) and the repository falls back to its id as directory name. Had the legacy syntax been resolved, it would have cloned into `repos/quando632` without any warning.
   This behaviour is also pinned by the unit test `RepositoryPropertiesTest.testLegacyCurlySyntaxIsNotResolved`.
7. Replace the `git_url` with the `@ask-variable(..)` example above and remove `GIT_USER` everywhere. On the first run you are asked `Please enter your github login: [«your-os-user»]` (the default comes from `$[USERNAME]`), the answer is written to `~/.ide/ide.properties`, and the fork is cloned. On a second run you are not asked again.

---

### Checklist for this PR

Make sure everything is checked before merging this PR. For further info please also see
our [DoD](https://github.com/devonfw/IDEasy/blob/main/documentation/contributing/DoD.adoc).

- [x] When running `mvn clean test` locally all tests pass and build is successful (`mvn verify -pl '!documentation'` → BUILD SUCCESS, `ide-cli`: 959 tests, 0 failures, 0 errors)
- [x] PR title is of the form `#«issue-id»: «brief summary»` (e.g. `#921: fixed setup.bat` and not `feature/921 fixed setup.bat`). If no issue ID exists, title only.
- [x] PR top-level comment summaries what has been done and contains link to addressed issue(s)
- [x] PR and issue(s) have suitable labels
- [x] Issue is set to `In Progress` and assigned to you *or* there is no issue (might happen for very small PRs)
- [x] You followed all [coding conventions](https://github.com/devonfw/IDEasy/blob/main/documentation/coding-conventions.adoc) (`mvn spotless:apply` and `mvn checkstyle:check` are clean for all changed files)
- [x] You have added the issue implemented by your PR in [CHANGELOG.adoc](https://github.com/devonfw/IDEasy/blob/main/CHANGELOG.adoc) unless issue is labelled with `internal`
- [x] You have not changed any dependency in `pom.xml` files or otherwise if runtime dependencies changed, you have updated our [LICENSE.asciidoc](https://github.com/devonfw/IDEasy/blob/main/documentation/LICENSE.adoc)
- [x] You have formulated clear instructions on how to test your contribution under "Testing instructions"

### CODE REVIEW

File: cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java [hunk 1]
```diff
@@ -88,20 +90,44 @@ protected void doRun() {
...
+   */
+  private RepositoryConfig prepareActiveRepositoryIgnoringInvalid(Path repositoryFile, boolean forceMode) {
 
-    RepositoryConfig config = RepositoryConfig.loadProperties(repositoryFile, this.context);
-    if (config == null) {
+    try {
+      return prepareActiveRepository(repositoryFile, forceMode);
+    } catch (CliAbortException e) {
+      throw e; // the user deliberately aborted - never swallow this
+    } catch (CliException e) {
```
Comment (JoelAdbu) [inline cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java:108]: We should also consider the special case where resolving a circular variable reference throws an IllegalStateException. Otherwise, one invalid repository could still terminate the entire batch setup.

```suggestion
} catch (CliAbortException | IllegalStateException e) {
```
This would fix the repository batch setup case. However, the root cause seems to be in AbstractEnvironmentVariables.resolveRecursive(): a circular variable reference is a user configuration error and should therefore probably result in a CliException, not an IllegalStateException. Catching IllegalStateException here could potentially hide unrelated programming errors from the same code path.

File: cli/src/test/java/com/devonfw/tools/ide/commandlet/RepositoryCommandletTest.java [hunk 2]
```diff
@@ -86,6 +86,63 @@ void testSetupAllRepositoriesInactive() {
...
+    good.setProperty("active", "true");
+    saveProperties(context, good);
+    RepositoryCommandlet rc = context.getCommandletManager().getCommandlet(RepositoryCommandlet.class);
+    // act
+    rc.run();
+    // assert
+    assertThat(context.getIdeHome().resolve(IdeContext.FOLDER_WORKSPACES).resolve(TEST_WORKSPACE).resolve(TEST_REPO)).isDirectory();
+    assertThat(context).logAtError().hasMessageContaining("Invalid template expression");
+  }
+
```
Comment (JoelAdbu) [inline cli/src/test/java/com/devonfw/tools/ide/commandlet/RepositoryCommandletTest.java:146]: We could also add a test to cover this edge case.

```suggestion
  @Test
  void testCircularVariableInOneRepositoryDoesNotPreventOtherRepositories() {
    IdeTestContext context = newContext(IdeContext.FOLDER_REPOSITORY);

    context.getVariables().getByType(EnvironmentVariablesType.CONF).set("REPO_A", "$[REPO_B]");
    context.getVariables().getByType(EnvironmentVariablesType.CONF).set("REPO_B", "$[REPO_A]");

    Properties broken = createDefaultProperties();
    broken.setProperty("active", "true");
    broken.setProperty("path", "broken-repo");
    broken.setProperty("git_url", "https://github.com/@ask-variable('REPO_A')/broken.git");
    saveProperties(context, broken, "broken.properties");

    Properties good = createDefaultProperties();
    good.setProperty("active", "true");
    saveProperties(context, good);

    RepositoryCommandlet rc = context.getCommandletManager().getCommandlet(RepositoryCommandlet.class);

    rc.run();

    Path goodRepository = context.getIdeHome()
        .resolve(IdeContext.FOLDER_WORKSPACES)
        .resolve(TEST_WORKSPACE)
        .resolve(TEST_REPO);

    assertThat(goodRepository).isDirectory();

  }

```

**Review by JoelAdbu — COMMENTED** [review]
Well done! I tested it locally and everything works fine. Please consider the minor suggestions below to ensure all edge cases are covered.

File: cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java (hunk 1 above)
Comment (quando632) [reply cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java:108]: I fixed it in the last commit.

Reproduced first: with `REPO_A=$[REPO_B]` / `REPO_B=$[REPO_A]` the run aborts with `IllegalStateException: Reached maximum recursion resolving $[REPO_A] ...` and the second, valid repository is not set up.

`resolveRecursive` now throws `CliException` for the maximum recursion. With that, `prepareActiveRepositoryIgnoringInvalid` needs no change at all the existing `catch (CliException)` covers it, and we avoid catching `IllegalStateException`, which is exactly the masking risk you described.

Checked that nothing relies on the old type: the only other `IllegalStateException` assertions in tests are the ambiguous-merge one in `XmlMergerTest:88` and a helper in `DirectoryMergerTest:155`, both different throw sites. Side effect is that a circular reference now surfaces as a clean CLI error in the configurator/merger too, instead of a stacktrace.

One note on the suggestion block itself: line 108 is the `catch (CliException e)`, so applying it literally gives `error: exception CliAbortException has already been caught` (the rethrow above already catches it) and would drop the `CliException` handling.

File: cli/src/test/java/com/devonfw/tools/ide/commandlet/RepositoryCommandletTest.java (hunk 2 above)
Comment (quando632) [reply cli/src/test/java/com/devonfw/tools/ide/commandlet/RepositoryCommandletTest.java:146]: Added as `testCircularVariableDoesNotPreventOtherRepositories`, close to your version.
Two changes: dropped the `@ask-variable('REPO_A')` wrapper, since a plain `$[REPO_A]` in the `git_url` triggers the same recursion more directly, and added an assertion on the logged error so the test also pins that the repository is reported rather than silently skipped.

File: cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java (hunk 1 above)
Comment (JoelAdbu) [reply cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java:108]: Thanks for fixing this. Regarding your note, you're correct, that was a typo on my side. Anyway, the fix in AbstractEnvironmentVariables was the right way to address it, so thank you.

Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java:108]: Just for the record. Switching the recursion error to `CliException` now prevents that a stracktrace gets printed by default (unless in debug mode). For bug-reports the stacktrace might still be helpful, but just a tiny detail not preventing merge.

Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java:108]: wouldn't our step infrastructure be much better here?
This seems to be the exact pattern we typically solve with steps what also gives better user-feedback.

File: cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryProperties.java [hunk 3]
```diff
@@ -47,6 +49,11 @@ final class RepositoryProperties {
 
   private final Properties properties;
 
+  private final IdeContext context;
```
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryProperties.java:52]: Nothing wrong but challenging the design:
It was already questionable before that `IdeContext` was passed to the constructor.
But adding it to this simple data object seems not a really good design.
Typically properties, configs, etc. should focus on their data and remain rather stupid objects not being connected to context and other state.
You could also pass the context as argument to get properties methods and resolve.
Or you could even keep the entire resolving out here leaving this as a stupid data object and add the resolving part to the method "transforming" these properties object to RepositoryConfig.

File: documentation/repository.adoc [hunk 4]
```diff
@@ -54,3 +54,26 @@ If omitted no build is triggered.
 |`link`|CSV of `«link»[=«path»]`|(optional) Comma separated list of symbolic links to link the repository into the workspace top-level folder. For a real-world use-case of this feature see link:ai.adoc[AI usage].
 |`active`|`true`|(optional) If set to `false` the repository is skipped during the link:setup.adoc[setup].
 |===
+
+== Variables and expressions
+
+Every value in a repository `.properties` file may contain link:variables.adoc[variables] using the `$[«VARIABLE»]` syntax as well as link:configurator.adoc[expressions] using the `@«function»(...)` syntax.
```
Comment (hohwille) [inline documentation/repository.adoc:60]:
```suggestion
Every value in a repository `.properties` file may contain link:variables.adoc[variables] using the `$[«VARIABLE»]` syntax as well as link:configurator.adoc#functions[expression functions] using the `@«function»(...)` syntax.
```

File: documentation/repository.adoc [hunk 5]
```diff
@@ -54,3 +54,26 @@ If omitted no build is triggered.
...
+path=IDEasy
+git_url=https://github.com/@ask-variable("GIT_USER", "Please enter your github login:", "home", "$[USERNAME]")/IDEasy.git
+git_remote=upstream:https://github.com/devonfw/IDEasy.git
+git_branch=main
+import=eclipse
+active=true
+```
+
+On the first setup the user is asked for the github login, the answer is stored as `GIT_USER` in `~/.ide/ide.properties` (`home`) and reused on every following run.
+The default `$[USERNAME]` is the login name of the current user on Windows - on Linux and macOS use `$[USER]` instead.
```
Comment (hohwille) [inline documentation/repository.adoc:75]: If you already found out this flaw that on Windows it is `$USER` and on Linux/mac it is `$USERNAME` then this approach is flawed since it is not working cross-platform.
IDEasy aims to provide a cross-platform solution.
Therefore we should not document a flaw but instead create an issue to solve the flaw.
I would suggest that we make USER variable also available on Windows and use it here instead of `$[USERNAME]`.

File: documentation/repository.adoc [hunk 6]
```diff
@@ -54,3 +54,26 @@ If omitted no build is triggered.
...
+
+NOTE: A repository with `active=false` is skipped before its properties are read, so a question from `@ask-variable` is never asked for a repository that is not set up.
+
+NOTE: The legacy `${«VARIABLE»}` syntax is intentionally *not* resolved here so that e.g. maven properties in a `build_cmd` are passed through untouched.
```
Comment (hohwille) [inline documentation/repository.adoc:79]: Please avoid too much documentation duplications in different places since this makes it hard to maintain.
Also this is not exactly true since there is a feature to enable this syntax for backward compatibility.

```suggestion
```

**Review by hohwille — APPROVED** [review]
@quando632 thanks for your PR. You properly implemented and tested the story well. 👍 
I left some comments for improvement. Please have a look.

**Review by hohwille — APPROVED** [review]
~~@quando632 thanks for your PR. You properly implemented and tested the story well. 👍 
I left some comments for improvement. Please have a look.~~

Review was submitted twice due to a GitHub error.

File: documentation/repository.adoc (hunk 5 above)
Comment (hohwille) [reply documentation/repository.adoc:75]: Acutally it is so trivial to fix this problem that it is not worth creating a new issue for it.
It is less effort if you just add `USER` to `IdeVariables` with fallback to USERNAME to make this work and fix the problem.

File: documentation/repository.adoc (hunk 6 above)
Comment (hohwille) [reply documentation/repository.adoc:79]: OK, got it. In your case, you are explicitly passing `false` so it is always disabled.
I understand your point but isn't crisp documentation more valuable?
If you say variables are supported via `$[...]` then why explicitly saying that `${...}` is not supported?
We already have long documentation and users tend to have little time to read all this.
IMHO we can still remove this to avoid information overload. There are so many things you could write that we are not doing or not supporting so lets better document what we support.

Comment (quando632) [reply documentation/repository.adoc]: I Agreed. Documenting what we support rather than what we do not is the better rule

File: documentation/repository.adoc (hunk 4 above)
Comment (quando632) [reply documentation/repository.adoc]: Applied. The anchor works, `=== Functions` is at `configurator.adoc:71`.

File: documentation/repository.adoc (hunk 5 above)
Comment (quando632) [reply documentation/repository.adoc]: Done. `USER` is in `IdeVariables` now with `USERNAME` as fallback, the example uses `$[USER]` and I added a row to `variables.adoc`.

Small thing on the direction: it is the other way round. Windows sets `USERNAME`, Linux and macOS set `USER`.

There was one catch. `EnvironmentVariablesSystem.getFlat(..)` blocks every variable registered in `IdeVariables.VARIABLES`, only `PATH`, `IDE_ROOT` and `HOME` are excluded. So just registering `USER` hid the real `USER` env variable on Linux and macOS, and the Windows fallback would have won everywhere. The test showed it immediately (`expected: "unix-login" but was: "windows-login"`). I added `USER` to that exception list, since it is an OS variable like `HOME` and `PATH`. Tell me if you prefer keeping the list at three entries and reading the env directly in the fallback instead.

File: cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java (hunk 1 above)
Comment (quando632) [reply cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java:108]: I checked this. Steps are clearly the pattern here. But there is a catch in this spot.

`Step.run(Runnable, boolean)` catches `RuntimeException | Error` and with `rethrow=false` it swallows everything, including `CliAbortException`. My rethrow is currently the only place in the production code that handles `CliAbortException` at all. With a plain step, a user who answers "no" to an `@ask-variable` question would get asked again for every following repository, and the `ProcessResult.ABORT` exit code would be lost.

Options I see:

1. keep the try/catch
2. use a step but keep the rethrow around it
3. make `Step.run` never swallow `CliAbortException`

I think 3 is the correct fix, but it touches shared code, so probably a separate issue. For now I would keep 1. What do you prefer?

File: cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryProperties.java (hunk 3 above)
Comment (quando632) [reply cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryProperties.java:52]: Fair point. The class uses the context for two things only: reading the file in the public constructor and `getVariables().resolve(..)`.

Moving the resolving into the transformation to `RepositoryConfig` does not work though. `RepositoryCommandlet` reads `isActive()` before the other properties on purpose, so an `@ask-variable` in a `git_url` never asks about a repository that gets skipped. If resolving happened only in the transformation, `active` would be read unresolved. The value cache is there for the same reason, resolving has side effects, so reading a property twice must not ask twice.

What about passing a `UnaryOperator<String>` resolver instead of the `IdeContext`? Then the data object only holds a function, no context and no state, and the caching and the ordering stay as they are. The test constructor would just pass `UnaryOperator.identity()`.

Does that work for you?

File: cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java (hunk 1 above)
Comment (quando632) [reply cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java:108]: Small correction: in this path there is no stacktrace at all, also not with `--debug`, because the handler only logs `e.getMessage()`. I can pass the exception to the log call if you want it for bug reports.

**Review by hohwille — APPROVED** [review]
@quando632 thanks for the update. Now perfect and ready for merge. 👍

File: cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryProperties.java (hunk 3 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryProperties.java:52]: It is already working as-is. It is just a small design flaw that is "not so nice".
I would suggest that we do not block this PR any further since everything is functionally correct.
Feel invited to create a issue or simply a PR for later cleanup.
We could even consider not supporting active property to use variables.
Also it would be possible to change the using code to do the transformation providing the context and then only working with the resulting config instead of using the properties object at all that is IMHO only an intermediate thing with the only purpose to encapsulate the parsing of the properties file.

File: cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java (hunk 1 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/git/repository/RepositoryCommandlet.java:108]: When we discussed the ability that a CliException can trigger rethrow automatically via the new boolean method, I also suggested that CliAbortException should do that just like CliFatalException since abort is thrown if the user decided to stop further processing.
Please also consider this for a cleanup issue / PR.

**Comment by quando632** [comment]
Retest on SNAPSHOT `2026.09.002-09_16_07-SNAPSHOT` on Windows 11 passed
